File: problems/cpp/226_invert_binary_tree.cpp
#include <iostream>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    void swap(TreeNode *node)
    {
        TreeNode *temp = node->left;
        node->left = node->right;
        node->right = temp;
    }

    TreeNode *invertTree(TreeNode *root)
    {
        if (root == nullptr)
            return root;

        invertTree(root->left);
        invertTree(root->right);

        swap(root);
        return root;
    }

    void PrintInOrder(TreeNode *root)
    {
        if (root == nullptr)
            return;

        std::cout << root->val << std::endl;

        PrintInOrder(root->left);
        PrintInOrder(root->right);
    }
};

int main()
{

    TreeNode left_23{6};
    TreeNode right_24{9};

    TreeNode left_21{1};
    TreeNode right_22{3};

    TreeNode left_1{2, &left_21, &right_22};
    TreeNode right_1{7, &left_23, &right_24};
    TreeNode tree{4, &left_1, &right_1};

    TreeNode *root_copy = &tree;
    Solution sol{};

    sol.PrintInOrder(&tree);
    std::cout << "-------------------------------" << std::endl;
    TreeNode *inverted_tree = sol.invertTree(&tree);
    sol.PrintInOrder(inverted_tree);

    return 0;
}

File: problems/python/226_invert_binary_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root == None:
            return None
        
        self.invertTree(root.right)
        self.invertTree(root.left)

        temp = root.left
        root.left = root.right
        root.right = temp

        return root
        
def print_tree(root):
    if root == None: 
        return

    print(root.val)

    print_tree(root.right)
    print_tree(root.left)
    
        
tree = TreeNode(4, TreeNode(2, TreeNode(1), TreeNode(3)), TreeNode(7, TreeNode(6), TreeNode(9)))

sol = Solution()

print_tree(tree)
sol.invertTree(tree)
print("---------------")
print_tree(tree)